haskell/test/Test/Ligo/BaseDAO/Plist.hs

```diff
@@ -0,0 +1,102 @@
+import Test.Tasty.Hedgehog (testProperty)
+
+import Hedgehog.Gen.Tezos.Crypto (genSecretKey)
+import Lorentz hiding (cast, concat, get, not)
+import Morley.Tezos.Address (mkKeyAddress)
+import Morley.Tezos.Crypto (toPublic)
+
+import Ligo.BaseDAO.Types
+import SMT.Model.BaseDAO.Plist
+import Test.Ligo.BaseDAO.Common (makeProposalKey, metadataSize)
+
+{-# ANN module ("HLint: ignore Reduce duplication" :: Text) #-}
+
+genProposalKeyList :: MonadGen m => Int -> m [ProposalKey]
+genProposalKeyList atLeast = do
+  i <- Gen.integral (Range.constant atLeast 30)
+  addrs <- genSecretKey
+    <&> (\secret -> mkKeyAddress . toPublic $ secret)
+    & Gen.list (Range.linear atLeast i)
+
+  pure $ (\addr ->
+      let proposalMeta = lPackValueRaw @Natural 1
+          metaSize = metadataSize proposalMeta
+          param = ProposeParams
+            { ppFrom = addr
+            , ppFrozenToken = metaSize
+            , ppProposalMetadata = proposalMeta
+            }
+      in makeProposalKey param
+    ) <$> addrs
+
+
+test_Plist :: [TestTree]
+test_Plist =
+  [ testProperty "Test conversion" $ withTests 50 $ property $ do
+      keyList <- forAll $ genProposalKeyList 0
+      let (pset, plist) = (OSet.fromList keyList, plistFromList keyList)
+
+      (toList pset) === (plistToList plist)
+
+  , testProperty "Test mem" $ withTests 50 $ property $ do
+      keyList <- forAll $ genProposalKeyList 1
+      let (pset, plist) = (OSet.fromList keyList, plistFromList keyList)
+
+      (i :: Int) <- forAll $ Gen.integral (Range.constant 0 (length keyList - 1))
+      let k = keyList !! i
+
+      (OSet.member k pset) === (plistMem k plist)
+
+    , testProperty "Test insert" $ withTests 50 $ property $ do
+      let (pset, plist) = (OSet.fromList [], plistFromList [])
+
+      keyList2 <- forAll $ genProposalKeyList 1
+      (i :: Int) <- forAll $ Gen.integral (Range.constant 0 (length keyList2 - 1))
+      let k = keyList2 !! i
+
+      (toList $ k <| pset)
+        === (plistToList $ plistInsert k plist)
+
+    , testProperty "Test delete" $ withTests 50 $ property $ do
+      keyList <- forAll $ genProposalKeyList 1
+      let (pset, plist) = (OSet.fromList keyList, plistFromList keyList)
+
+      (i :: Int) <- forAll $ Gen.integral (Range.constant 0 (length keyList - 1))
+      let k = keyList !! i
+
+      (toList $ OSet.delete k pset) === (plistToList $ plistDelete k plist)
+
+    , testProperty "Test updateHead" $ withTests 3 $ property $ do
+      keyList <- forAll $ genProposalKeyList 1
+      let plist = plistFromList keyList
+
+      (i :: Int) <- forAll $ Gen.integral (Range.constant 0 (length keyList - 1))
+      let k = keyList !! i
+
+      -- It's much easier to test updateHead f against a list.
+      (listUpdateHead k keyList) === (plistToList $ plistUpdateHead k plist)
+  ]
+    where
+      listUpdateHead k l =
+        case l of
+          (x:xs) ->
+            if (x == k) then l
+            else listUpdateHead k xs
+          _ -> []
```

Comment:
Really not sure how to do this, also the main operations that modify the list are only `add` and `delete`, so I'm not sure if this is worth doing or not.

Reply:
>so I'm not sure if this is worth doing or not.

May be I am missing something, but in implementations like this, bug are often hidden in some specific combination of operations. For example, add 5 items, delete 5 items, and then add another item. Normally we don't have to write tests for them because we don't usually have to implement core data structures ourself.

But since we are doing that here, and we are doing that in a smart contract for something that is part of its core operation, I think we should spare no efforts in testing this implementation.  @pasqu4le please correct me if I am wrong here.

>Really not sure how to do this,

I was thinking something like this.  Write a small contract with storage type `plist`, and entrypoints to load it, then entrypoints to  add/delete from it. With this, we can implement the property tests by generating a list of operations, and implementing the tests the same way we implement SMT tests for the main contract.

Reply:
If this makes sense, I am ok with adding the tests in another issue/PR.

Reply:
If needed, I'm good with creating another issue for it as well.

Reply:
@sras Created the issue: https://github.com/tezos-commons/baseDAO/issues/319, let me know if there are more changes you want to be made.

Reply:
I agree, normally the current tests would be enough IMO, but given that is based on a new (fairly complex) data structure we should test this more carefully.

I think it was a good idea to merge this PR now, but let's not put these tests off for much longer
@rinn7e I assigned the new issue to you